File: projects/relay/engine-python/src/relay_engine/rng.py
"""Portable seeded randomness.

Byte-for-byte the same algorithm as LUDO's dice and ALIBI's deal (splitmix64
seeding + xorshift64*). Repeated here rather than shared because `shared/` holds
contracts and data only — never executable code — and because the Java engine
must reproduce every draw exactly or the conformance vectors of ADR-0002 are
worthless.

    state  = splitmix64(seed)          # scrambles weak seeds like 0, 1, 2
    x     ^= x >> 12                   # xorshift64*
    x     ^= x << 25
    x     ^= x >> 27
    value  = x * 0x2545F4914F6CDD1D
    pick   = (value >> 33) % n

All arithmetic is unsigned 64-bit, discarding overflow.

The shuffle is Fisher-Yates driven by `below`, iterating from the last index
down to 1 and drawing `below(i + 1)` each step. That iteration order is part of
the spec: a Java port that loops upward consumes the same draws and produces a
different track. `sample` shuffles a copy of the WHOLE pool and takes the first
k — so it consumes len(pool) - 1 draws, not k. Also spec.
"""

from __future__ import annotations

MASK64 = 0xFFFFFFFFFFFFFFFF
MULTIPLIER = 0x2545F4914F6CDD1D


def splitmix64(seed: int) -> int:
    z = (seed + 0x9E3779B97F4A7C15) & MASK64
    z = ((z ^ (z >> 30)) * 0xBF58476D1CE4E5B9) & MASK64
    z = ((z ^ (z >> 27)) * 0x94D049BB133111EB) & MASK64
    return (z ^ (z >> 31)) & MASK64


class Rng:
    """A seeded source of small uniform picks."""

    def __init__(self, seed: int) -> None:
        self.seed = seed
        state = splitmix64(seed)
        # xorshift64* cannot recover from an all-zero state.
        self._state = state or 0x9E3779B97F4A7C15

    def below(self, n: int) -> int:
        """Uniform integer in [0, n)."""
        if n <= 0:
            raise ValueError("n must be positive")
        x = self._state
        x ^= x >> 12
        x = (x ^ (x << 25)) & MASK64
        x ^= x >> 27
        self._state = x
        return (((x * MULTIPLIER) & MASK64) >> 33) % n

    def between(self, low: int, high: int) -> int:
        """Uniform integer in [low, high]."""
        return low + self.below(high - low + 1)

    def shuffle(self, items: list) -> None:
        """In-place Fisher-Yates, high index down — iteration order is spec."""
        for i in range(len(items) - 1, 0, -1):
            j = self.below(i + 1)
            items[i], items[j] = items[j], items[i]

    def sample(self, items: list, k: int) -> list:
        """First k of a shuffled copy. Order of the result is the shuffle's."""
        pool = list(items)
        self.shuffle(pool)
        return pool[:k]

File: projects/relay/engine-java/src/main/java/com/llmeval/relay/engine/Rng.java
package com.llmeval.relay.engine;

import java.util.ArrayList;
import java.util.List;

/**
 * Portable seeded randomness — splitmix64 seeding over xorshift64*.
 *
 * <p>Must reproduce {@code relay_engine/rng.py} draw for draw. Java's {@code long} is signed and
 * Python's ints are unbounded, so every shift here is the <em>unsigned</em> {@code >>>}: a signed
 * {@code >>} would sign-extend and the two engines would diverge on roughly half of all states,
 * silently, and only for some seeds.
 *
 * <p>Two iteration orders are part of the spec rather than implementation detail. {@link #shuffle}
 * runs from the last index down to 1. {@link #sample} shuffles a copy of the <em>whole</em> pool
 * and takes the first k, consuming {@code pool.size() - 1} draws rather than k — a port that drew
 * k times would produce a different track from the same seed.
 */
public final class Rng {

    private static final long MULTIPLIER = 0x2545F4914F6CDD1DL;

    private final long seed;
    private long state;

    public Rng(long seed) {
        this.seed = seed;
        long start = splitmix64(seed);
        // xorshift64* cannot recover from an all-zero state.
        this.state = start == 0 ? 0x9E3779B97F4A7C15L : start;
    }

    public long seed() {
        return seed;
    }

    static long splitmix64(long seed) {
        long z = seed + 0x9E3779B97F4A7C15L;
        z = (z ^ (z >>> 30)) * 0xBF58476D1CE4E5B9L;
        z = (z ^ (z >>> 27)) * 0x94D049BB133111EBL;
        return z ^ (z >>> 31);
    }

    /** Uniform integer in [0, n). */
    public int below(int n) {
        if (n <= 0) {
            throw new IllegalArgumentException("n must be positive");
        }
        long x = state;
        x ^= x >>> 12;
        x ^= x << 25;
        x ^= x >>> 27;
        state = x;
        // >>> 33 leaves 31 bits, so the value is non-negative and the modulo is plain int work.
        return (int) (((x * MULTIPLIER) >>> 33) % n);
    }

    /** Uniform integer in [low, high]. */
    public int between(int low, int high) {
        return low + below(high - low + 1);
    }

    /** In-place Fisher-Yates, high index down — iteration order is spec. */
    public <T> void shuffle(List<T> items) {
        for (int i = items.size() - 1; i > 0; i--) {
            int j = below(i + 1);
            T swap = items.get(i);
            items.set(i, items.get(j));
            items.set(j, swap);
        }
    }

    /** First k of a shuffled copy. Order of the result is the shuffle's. */
    public <T> List<T> sample(List<T> items, int k) {
        List<T> pool = new ArrayList<>(items);
        shuffle(pool);
        return new ArrayList<>(pool.subList(0, k));
    }
}
